Reactive Implementation of AuthorizedClientServiceOAuth2AuthorizedCli…

ankurpathak opened the pull request:
…entManager

ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager is reactive
version of AuthorizedClientServiceOAuth2AuthorizedClientManager

Fixes: gh-7569

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java:
```diff
@@ -0,0 +1,136 @@
+ * @see ReactiveOAuth2AuthorizedClientService
+ * @since 5.3
+ */
+public final class ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager implements ReactiveOAuth2AuthorizedClientManager {
```
Let's rename to `OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager`

ankurpathak replied:
Resolved

jgrandja replied:
Apologies. I meant to say `AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager` - without `OAuth2` prefix so the naming is consistent with existing `AuthorizedClientServiceOAuth2AuthorizedClientManager`

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java:
```diff
@@ -0,0 +1,136 @@
+ * @since 5.3
+ */
+public final class ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager implements ReactiveOAuth2AuthorizedClientManager {
+	private final ReactiveClientRegistrationRepository reactiveClientRegistrationRepository;
```
Please remove `reactive` prefix for instance variables. See `DefaultReactiveOAuth2AuthorizedClientManager` for naming conventions.

ankurpathak replied:
Resolved

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java:
```diff
@@ -0,0 +1,136 @@
+	private final ReactiveClientRegistrationRepository reactiveClientRegistrationRepository;
+	private final ReactiveOAuth2AuthorizedClientService reactiveAuthorizedClientService;
+	private ReactiveOAuth2AuthorizedClientProvider reactiveAuthorizedClientProvider = context -> null;
+	private Function<OAuth2AuthorizeRequest, Map<String, Object>> contextAttributesMapper = new DefaultContextAttributesMapper();
```
The type should be `Function<OAuth2AuthorizeRequest, Mono<Map<String, Object>>>` (return `Mono`). See `DefaultReactiveOAuth2AuthorizedClientManager.contextAttributesMapper`

ankurpathak replied:
Resolved

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java:
```diff
@@ -0,0 +1,136 @@
+public final class ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager implements ReactiveOAuth2AuthorizedClientManager {
+	private final ReactiveClientRegistrationRepository reactiveClientRegistrationRepository;
+	private final ReactiveOAuth2AuthorizedClientService reactiveAuthorizedClientService;
+	private ReactiveOAuth2AuthorizedClientProvider reactiveAuthorizedClientProvider = context -> null;
```
Update to `context -> Mono.empty()`

ankurpathak replied:
Resolved

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java:
```diff
@@ -0,0 +1,136 @@
+	 *
+	 * @param reactiveAuthorizedClientProvider the {@link ReactiveOAuth2AuthorizedClientProvider} used for authorizing (or re-authorizing) an OAuth 2.0 Client
+	 */
+	public void setReactiveOAuth2AuthorizedClientProvider(ReactiveOAuth2AuthorizedClientProvider reactiveAuthorizedClientProvider) {
```
Rename to `setAuthorizedClientProvider`

ankurpathak replied:
Resolved

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java:
```diff
@@ -0,0 +1,136 @@
+											})
+											.build()
+				).flatMap(authorizationContext -> this.reactiveAuthorizedClientProvider.authorize(authorizationContext)
+						.doOnNext(x -> reactiveAuthorizedClientService.saveAuthorizedClient(x, principal))
```
Rename `x` to `authorizedClient`

ankurpathak replied:
Renamed to _authorizedClient as conflicting with local variable authorizedClient.

ghostd commented:
Hi,

(not sure where i should post this comment... so i put it here)

I had to change some code of the class to make it work with my use case.

# How i tested this PR (maybe wrongly?)

(in my code, i already renamed `ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager` to `OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager`)
```java
    ReactiveClientRegistrationRepository clientRegistrationRepository = new InMemoryReactiveClientRegistrationRepository(clientRegistration);
    ReactiveOAuth2AuthorizedClientService authorizedClientService = new InMemoryReactiveOAuth2AuthorizedClientService(clientRegistrationRepository);

    OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager authorizedClientManager =
        new OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager(
                clientRegistrationRepository,
                authorizedClientService
        );
    authorizedClientManager.setReactiveOAuth2AuthorizedClientProvider(new SpecificPartnerReactiveOAuth2AuthorizedClientProvider());

```

The current implementation does not save the authorized client into the repository.

# The code i changed to make it work

I changed
```java
.doOnNext(x -> reactiveAuthorizedClientService.saveAuthorizedClient(x, principal))
```
to
```java
.flatMap(authClient -> reactiveAuthorizedClientService.saveAuthorizedClient(authClient, principal).thenReturn(authClient))
```

Regards

ankurpathak commented:
> Hi,
> 
> (not sure where i should post this comment... so i put it here)
> 
> I had to change some code of the class to make it work with my use case.
> 
> # How i tested this PR (maybe wrongly?)
> (in my code, i already renamed `ReactiveOAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager` to `OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager`)
> 
> ```java
>     ReactiveClientRegistrationRepository clientRegistrationRepository = new InMemoryReactiveClientRegistrationRepository(clientRegistration);
>     ReactiveOAuth2AuthorizedClientService authorizedClientService = new InMemoryReactiveOAuth2AuthorizedClientService(clientRegistrationRepository);
> 
>     OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager authorizedClientManager =
>         new OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager(
>                 clientRegistrationRepository,
>                 authorizedClientService
>         );
>     authorizedClientManager.setReactiveOAuth2AuthorizedClientProvider(new SpecificPartnerReactiveOAuth2AuthorizedClientProvider());
> ```
> 
> The current implementation does not save the authorized client into the repository.
> 
> # The code i changed to make it work
> I changed
> 
> ```java
> .doOnNext(x -> reactiveAuthorizedClientService.saveAuthorizedClient(x, principal))
> ```
> 
> to
> 
> ```java
> .flatMap(authClient -> reactiveAuthorizedClientService.saveAuthorizedClient(authClient, principal).thenReturn(authClient))
> ```
> 
> Regards

@ghostd I will wait fot @jgrandja to review above commets ad if required will do the needful.

ghostd commented:
Hi,

I did some tests with the new version. It seems if `DefaultContextAttributesMapper#apply` returns (a mono of) an empty map, the authentication does not occur. If i change the DefaultContextAttributesMapper to return a non-empty map, the authentication succeeds.

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java:
```diff
@@ -0,0 +1,147 @@
+									attributes.putAll(contextAttributes);
+								}).build())
+				).flatMap(authorizationContext -> this.authorizedClientProvider.authorize(authorizationContext)
+						.doOnNext(_authorizedClient -> authorizedClientService.saveAuthorizedClient(_authorizedClient, principal))
```
@ghostd Thanks for catching this...

> `.flatMap(authClient -> reactiveAuthorizedClientService.saveAuthorizedClient(authClient, principal).thenReturn(authClient))`

@ankurpathak This bug was fixed in this [commit](https://github.com/spring-projects/spring-security/commit/8584b12c8da9b73b2579df91b81ff86a49e4daa0#diff-b52f0a944955db35cc1072fdbb4393d2R108).

Can you also apply the update to the test to make use of `PublisherProbe<Void> saveAuthorizedClientProbe` to ensure we cover the scenario of the Mono being subscribed to. See the commit that adds this additional test

jgrandja commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager.java:
```diff
@@ -0,0 +1,147 @@
+
+		@Override
+		public Mono<Map<String, Object>> apply(OAuth2AuthorizeRequest authorizeRequest) {
+			ServerWebExchange serverWebExchange = authorizeRequest.getAttribute(ServerWebExchange.class.getName());
```
`ServerWebExchange` won't be available in this scenario so `.defaultIfEmpty(Collections.emptyMap())` will execute every time. The logic should be similar to `AuthorizedClientServiceOAuth2AuthorizedClientManager.DefaultContextAttributesMapper`. Looks like a test is missing for this scenario, can you please add.

jgrandja replied:
Also see this [comment](https://github.com/spring-projects/spring-security/pull/7589#issuecomment-559243184) as the logic for `.flatMap(contextBuilder -> this.contextAttributesMapper.apply(authorizeRequest)` is incorrect. Please add a test for this as well. If `contextAttributesMapper` returns an empty `Map`, `authorizedClientProvider.authorize()` should still be called.

philsttr commented:
Any chance this could be added in 5.2.x ?

philsttr commented on oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/OAuth2AuthorizedClientServiceReactiveOAuth2AuthorizedClientManagerTests.java:
```diff
@@ -0,0 +1,300 @@
+				.build();
+		Mono<OAuth2AuthorizedClient> authorizedClient = this.authorizedClientManager.authorize(authorizeRequest);
+
+		authorizedClient.subscribe();
```
Don't `subscribe()` here.  Instead, rely on `StepVerifier` subscribing (during `verifyComplete()`).   Move the verifications and assertions after the StepVerifier.

The same comment applies to all of your tests where you explicitly call .subscribe()

ankurpathak commented:
@jgrandja you can takeover this as I tried a lo on this but can't make it.

philsttr commented:
I can pick this up to continue it.

jgrandja commented:
@ankurpathak No worries and thank you for your work thus far.

@philsttr That would be great if you can take over this. Much appreciated.

philsttr commented:
Followup PR created as #7702 

Please close this PR (#7589) in favor of #7702

jgrandja commented:
Closing in favour of #7702
